fix(cli): stale process

kobenguyent opened the pull request:
## Motivation/Description of the PR
- Resolves #4197
- resolves #4358 
- resolves #4211

```
  mocha:runner _addEventListener(): adding for event uncaughtException; 0 current listeners +8ms
  mocha:runner _addEventListener(): adding for event unhandledRejection; 0 current listeners +0ms
  mocha:runner run(): starting +2ms
  mocha:runner run(): emitting start +0ms

  mocha:runner run(): emitted start +1ms
  mocha:runner runSuite(): running  +0ms
  mocha:runner runSuite(): running Error 1 @error1: +1ms
Error 1 @error1 --
  mocha:runner _addEventListener(): adding for event error; 0 current listeners +1ms
    [1]  Starting recording promises
    Retries: 3
    Timeouts: 
 › [Session] Starting singleton browser session
  test issue
  mocha:runner _addEventListener(): adding for event error; 0 current listeners +1s
  mocha:runner _addEventListener(): adding for event error; 0 current listeners +1s
    I am on page "http://example.org"
    › [Browser:Error] Failed to load resource: the server responded with a status of 404 (Not Found)
    I wait for visible ".nothing", 1
    [1]  Error (Non-Terminated) | Error: element (.nothing) still not visible after 1 sec
locator.waitFor: Timeout 1000ms exceeded.
Call log:
  - waiting for locator('.nothing').first() to be visible
 | (err) => { step.status = 'failed'; step.endTime = ...
    [1] Error | Error: element (.nothing) still not visible after 1 sec
locator.waitFor: Timeout 1000ms exceeded.
Call log:
  - waiting for locator('.nothing').first() to be visible
 (e) => { const err = (recorder.getAsyncErr() === n...
    [1] <teardown>  Stopping recording promises
 › <screenshotOnFail> Test failed, try to save a screenshot
 › Screenshot is saving to /Users/t/Desktop/projects/codeceptjs-debug/output/test_issue.failed.png
  mocha:runner _addEventListener(): adding for event error; 0 current listeners +2s
    [2]  Starting recording promises
  mocha:runner _addEventListener(): adding for event error; 0 current listeners +35ms
  mocha:runner runSuite(): running Error 2 @error2: +2ms
  mocha:runner runSuite(): bailing +0ms
  mocha:runner runSuite(): running Error 3 @error3: +0ms
  mocha:runner runSuite(): bailing +0ms
  mocha:runner runSuite(): running Error 4 @error4: +0ms
  mocha:runner runSuite(): bailing +0ms
  mocha:runner runSuite(): running Working @working: +1ms
  mocha:runner runSuite(): bailing +0ms
  mocha:runner run(): root suite completed; emitting end +0ms

  OK  | 0 passed   // 4s
  mocha:runner run(): emitted end +1ms

```

From @Horsty80 

- When a throw error occurs inside promise, that stale the process, by rewrite with good reject and done call, we prevent this to happen
- Resolves #4367 

- maybe this pr can help this issue #4358

## Type of change
- [ ] :bug: Bug fix


## Checklist:

- [ ] Tests have been added
- [ ] Documentation has been added (Run `npm run docs`)
- [ ] Lint checking (Run `npm run lint`)
- [ ] Local tests are passed (Run `npm test`)

kobenguyent committed:
fix(cli): stale process

kobenguyent commented:
ready for testing: **codeceptjs@3.6.3-beta.2**

Horsty80 commented on lib/cli.js:
```diff
@@ -186,20 +186,24 @@ class Cli extends Base {
+          test.err = err;
+          return test;
+        } catch (e) {
+          throw Error(e);
```
**question:**
@kobenguyent throwing an error are you sure about it ? i mean the issue about stale process is, i think, throw error, so maybe just log the error and return something else maybe ?

kobenguyent replied:
I think we should throw an error to stop the process here.

Horsty80 replied:
Ok i see, so the issue isn't that this code throw an error but is that in case of error nothing is throw and the process didn't get the trigger to stop ?
Seems legit indeed :)
Thanks, i'm trying this version right now, but without retry so i didn't confirm that modif is working

kobenguyent replied:
I tried the patch using the example here. https://github.com/codeceptjs/CodeceptJS/issues/4197 looking good so far.

Horsty80 committed:
fix: handle throw error inside retryTo promise (#4377)

* fix(webApi): error message of dontSeeCookie (#4357)

* Fixed error message for dontSeeCookie()

* fix(cli): gherkin command init with TypeScript (#4366)

* DOC: Update contributor faces

* fix(doc): Expect helper is not described correctly (#4370)

* fix: Puppeteer helper doc typo (#4369)

* chore(deps): bump devtools from 8.36.1 to 8.38.0 (#4374)

* chore(deps-dev): bump playwright from 1.44.0 to 1.44.1 (#4376)

Bumps [playwright](https://github.com/microsoft/playwright) from 1.44.0 to 1.44.1.
- [Release notes](https://github.com/microsoft/playwright/releases)
- [Commits](https://github.com/microsoft/playwright/compare/v1.44.0...v1.44.1)

---
updated-dependencies:
- dependency-name: playwright
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* chore(deps): bump axios from 1.6.7 to 1.7.2 (#4372)

* chore(deps-dev): bump typedoc-plugin-markdown from 3.17.1 to 4.0.3 (#4373)

* fix: handle throw error inside retryTo promise

* fix: test due to update on retryTo plugin

* fix: add a test case succeed after 2 retry

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: Arthur Stankevich <a-stankevich@users.noreply.github.com>
Co-authored-by: Ando NARY <36852616+andonary@users.noreply.github.com>
Co-authored-by: GitHub Action <action@github.com>
Co-authored-by: KobeN <7845001+kobenguyent@users.noreply.github.com>
Co-authored-by: Yoann Fleury <yoann.fleury@yahoo.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

kobenguyent committed:
resolve conflicts

Horsty80 committed:
Ensure scenario termination (#4378)

* chore: Add test:stale script for running state_test.js

* fix: handle async function better to prevent stale the process

* fix: done() call multiple times

* chore: rename files and test description

* fix: playwright error on test

* fix: stale process due to retryTo error handler

* fix: lint previous code

* fix: error due to the previous merge

* fix: handle async function better to prevent stale the process

* fix: lint previous code

---------

Co-authored-by: KobeN <7845001+kobenguyent@users.noreply.github.com>

kobenguyent commented:
ready for testing: codeceptjs@3.6.3-beta.3

DavertMik reviewed (approved):
Good work!
Looks good to me. Seems like a valid solution
